File: ohara-testing-util/src/main/scala/com/island/ohara/integration/OharaTestUtil.scala
package com.island.ohara.integration

import java.util.concurrent.TimeUnit

import com.island.ohara.client.ConnectorClient
import com.island.ohara.client.util.CloseOnce
import com.island.ohara.client.util.CloseOnce._
import com.island.ohara.common.util.CommonUtil

import scala.concurrent.duration._

/**
  * This class create a kafka services having 1 zk instance and 1 broker default. Also, this class have many helper methods to make
  * test more friendly.
  *
  * How to use this class:
  * 1) create the OharaTestUtil with 1 broker (you can assign arbitrary number from brokers)
  * val testUtil = OharaTestUtil.localBrokers(1)
  * 2) get the basic|producer|consumer OharaConfiguration
  * val config = testUtil.producerConfig
  * 3) instantiate your producer or consumer
  * val producer = new KafkaProducer<Array<Byte>, Array<Byte>>(config, new ByteArraySerializer, new ByteArraySerializer)
  * 4) do what you want for your producer and consumer
  * ...
  * 5) close OharaTestUtil
  * testUtil.close()
  *
  * see TestOharaTestUtil for more examples
  * NOTED: the close() will shutdown all services including the passed consumers (see run())
  *
  */
class OharaTestUtil private[integration] (zk: Zookeepers, brokers: Brokers, workers: Workers) extends CloseOnce {
  private[this] var localDb: Database = _
  private[this] var _connectorClient: ConnectorClient = _
  private[this] var localFtpServer: FtpServer = _
  private[this] var localHdfs: Hdfs = _

  /**
    * Exposing the brokers connection. This list should be in the form <code>host1:port1,host2:port2,...</code>.
    *
    * @return brokers connection information
    */
  def brokersConnProps: String = {
    Option(brokers)
      .map(_.connectionProps)
      .getOrElse(throw new RuntimeException(
        s"Brokers do not exist. Because Workers exist in supply environment, then we don't create embedded Brokers. Please do not operate it"))
  }

  /**
    * Exposing the workers connection. This list should be in the form <code>host1:port1,host2:port2,...</code>.
    *
    * @return workers connection information
    */
  def workersConnProps: String = {
    Option(workers).map(_.connectionProps).getOrElse(throw new RuntimeException(s"Workers do not exist"))
  }

  def connectorClient: ConnectorClient = {
    // throw exception if there is no worker cluster
    if (_connectorClient == null) _connectorClient = ConnectorClient(workers.connectionProps)
    _connectorClient
  }

  def hdfs: Hdfs = {
    if (localHdfs == null) localHdfs = Hdfs()
    localHdfs
  }

  def dataBase: Database = {
    if (localDb == null) localDb = Database()
    localDb
  }

  def ftpServer: FtpServer = {
    if (localFtpServer == null) localFtpServer = FtpServer()
    localFtpServer
  }

  override protected def doClose(): Unit = {
    CloseOnce.close(_connectorClient)
    CloseOnce.close(localDb)
    CloseOnce.close(localFtpServer)
    CloseOnce.close(hdfs)
    CloseOnce.close(workers)
    CloseOnce.close(brokers)
    CloseOnce.close(zk)
  }

}

object OharaTestUtil {

  /**
    * helper method. Loop the specified method until timeout or get true from method
    *
    * @param f            function
    * @param d            duration
    * @param freq         frequency to call the method
    * @param useException true make this method throw exception after timeout.
    * @return false if timeout and (useException = true). Otherwise, the return value is true
    */
  def await(f: () => Boolean, d: Duration, freq: Duration = 500 millis, useException: Boolean = true): Boolean = {
    val startTs = CommonUtil.current()
    while (d.toMillis >= (System.currentTimeMillis() - startTs)) {
      if (f()) return true
      else TimeUnit.MILLISECONDS.sleep(freq.toMillis)
    }
    if (useException) throw new IllegalStateException("timeout") else false
  }

  /**
    * Create a test util with multi-brokers.
    * NOTED: don't call the worker and hdfs service. otherwise you will get exception
    *
    * @return a test util
    */
  def brokers(): OharaTestUtil = {
    var zk: Zookeepers = null
    val brokers = Brokers {
      if (zk == null) zk = Zookeepers()
      zk
    }
    new OharaTestUtil(zk, brokers, null)
  }

  /**
    * Create a test util with multi-brokers and multi-workers.
    * NOTED: don't call the hdfs service. otherwise you will get exception
    *
    * @return a test util
    */
  def workers(): OharaTestUtil = {
    var zk: Zookeepers = null
    var brokers: Brokers = null
    val workers = Workers {
      if (brokers == null) brokers = Brokers {
        if (zk == null) zk = Zookeepers()
        zk
      }
      brokers
    }
    new OharaTestUtil(zk, brokers, workers)
  }

  /**
    * Create a test util with local file system.
    * NOTED: don't call the workers and brokers service. otherwise you will get exception
    *
    * @return a test util
    */
  def localHDFS(): OharaTestUtil = new OharaTestUtil(null, null, null)

  val HELP_KEY = "--help"
  val TTL_KEY = "--ttl"
  val USAGE = s"[Usage] $TTL_KEY"

  def main(args: Array[String]): Unit = {
    if (args.length == 1 && args(0) == HELP_KEY) {
      println(USAGE)
      return
    }
    if (args.length % 2 != 0) throw new IllegalArgumentException(USAGE)
    var ttl = 9999
    args.sliding(2, 2).foreach {
      case Array(TTL_KEY, value) => ttl = value.toInt
      case _                     => throw new IllegalArgumentException(USAGE)
    }
    doClose(OharaTestUtil.workers()) { util =>
      println("wait for the mini kafka cluster")
      TimeUnit.SECONDS.sleep(5)
      println(s"Succeed to run the mini brokers: ${util.brokersConnProps} and workers:${util.workersConnProps}")
      println(
        s"enter ctrl+c to terminate the mini broker cluster (or the cluster will be terminated after $ttl seconds")
      TimeUnit.SECONDS.sleep(ttl)
    }
  }
}

File: ohara-testing-util/src/main/java/com/island/ohara/integration/OharaTestUtil.java
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.integration;

import com.island.ohara.common.util.Releasable;
import com.island.ohara.common.util.ReleaseOnce;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

/**
 * This class create a kafka services having 1 zk instance and 1 broker default. Also, this class
 * have many helper methods to make test more friendly.
 *
 * <p>How to use this class: 1) create the OharaTestUtil with 1 broker (you can assign arbitrary
 * number from brokers) val testUtil = OharaTestUtil.localBrokers(1) 2) get the
 * basic|producer|consumer OharaConfiguration val config = testUtil.producerConfig 3) instantiate
 * your producer or consumer val producer = new KafkaProducer<Array<Byte>, Array<Byte>>(config, new
 * ByteArraySerializer, new ByteArraySerializer) 4) do what you want for your producer and consumer
 * ... 5) close OharaTestUtil testUtil.close()
 *
 * <p>see TestOharaTestUtil for more examples NOTED: the close() will shutdown all services
 * including the passed consumers (see run())
 */
public class OharaTestUtil extends ReleaseOnce {
  private Database localDb;
  private FtpServer localFtpServer;
  private Hdfs localHdfs;
  private final Zookeepers zk;
  private final Brokers brokers;
  private final Workers workers;

  private OharaTestUtil(Zookeepers zk, Brokers brokers, Workers workers) {
    this.zk = zk;
    this.brokers = brokers;
    this.workers = workers;
  }

  /**
   * Exposing the brokers connection. This list should be in the form <code>
   * host1:port1,host2:port2,...</code>.
   *
   * @return brokers connection information
   */
  public String brokersConnProps() {
    return Optional.ofNullable(brokers)
        .map(Brokers::connectionProps)
        .orElseThrow(
            () ->
                new RuntimeException(
                    "Brokers do not exist. Because Workers exist in supply environment, then we don't create embedded Brokers. Please do not operate it"));
  }

  /**
   * Exposing the workers connection. This list should be in the form <code>
   * host1:port1,host2:port2,...</code>.
   *
   * @return workers connection information
   */
  public String workersConnProps() {
    return Optional.ofNullable(workers)
        .map(Workers::connectionProps)
        .orElseThrow(() -> new RuntimeException("Workers do not exist"));
  }

  public Hdfs hdfs() {
    if (localHdfs == null) localHdfs = Hdfs.of();
    return localHdfs;
  }

  public Database dataBase() {
    if (localDb == null) localDb = Database.of();
    return localDb;
  }

  public FtpServer ftpServer() {
    if (localFtpServer == null) localFtpServer = FtpServer.of();
    return localFtpServer;
  }

  @Override
  protected void doClose() {
    Releasable.close(localDb);
    Releasable.close(localFtpServer);
    Releasable.close(localHdfs);
    Releasable.close(workers);
    Releasable.close(brokers);
    Releasable.close(zk);
  }

  /**
   * create a test util with a broker cluster based on single node. NOTED: don't call the worker and
   * hdfs service. otherwise you will get exception
   *
   * @return a test util
   */
  public static OharaTestUtil broker() {
    return brokers(1);
  }

  /**
   * Create a test util with multi-brokers. NOTED: don't call the worker and hdfs service. otherwise
   * you will get exception
   *
   * @return a test util
   */
  public static OharaTestUtil brokers(int numberOfBrokers) {
    AtomicReference<Zookeepers> zk = new AtomicReference<>();
    Brokers brokers =
        Brokers.of(
            () -> {
              if (zk.get() == null) {
                zk.set(Zookeepers.of());
              }
              return zk.get();
            },
            numberOfBrokers);
    return new OharaTestUtil(zk.get(), brokers, null);
  }

  /**
   * create a test util with a worker/broker cluster based on single node. NOTED: don't call the
   * worker and hdfs service. otherwise you will get exception
   *
   * @return a test util
   */
  public static OharaTestUtil worker() {
    return workers(1);
  }

  /**
   * Create a test util with multi-brokers and multi-workers. NOTED: don't call the hdfs service.
   * otherwise you will get exception.
   *
   * <p>NOTED: the default number of brokers is 3
   *
   * @return a test util
   */
  public static OharaTestUtil workers(int numberOfWorkers) {
    AtomicReference<Zookeepers> zk = new AtomicReference<>();
    AtomicReference<Brokers> brokers = new AtomicReference<>();

    Workers workers =
        Workers.of(
            () -> {
              if (brokers.get() == null) {
                brokers.set(
                    Brokers.of(
                        () -> {
                          if (zk.get() == null) zk.set(Zookeepers.of());
                          return zk.get();
                        },
                        numberOfWorkers));
              }
              return brokers.get();
            },
            numberOfWorkers);
    return new OharaTestUtil(zk.get(), brokers.get(), workers);
  }

  /**
   * Create a test util with local file system. NOTED: don't call the workers and brokers service.
   * otherwise you will get exception
   *
   * @return a test util
   */
  public static OharaTestUtil localHDFS() {
    return new OharaTestUtil(null, null, null);
  }
}
